internal/docs/design.md

```diff
@@ -184,6 +187,102 @@ needed). However, this would result in inconsistent names over time (e.g., some
+
+To enable the [Backing services factor][] of a Twelve-Factor Application, Go
+Cloud includes the ability to construct each of its API objects using
+identifying URLs. The concrete type should have API like this:
+
+```go
+package foo
+
+// A type that implements WidgetURLOpener can open widgets based on a URL.
+// The opener must not modify the URL argument. OpenWidgetURL must be safe to
+// call from multiple goroutines.
+//
+// For maximum flexibility, WidgetURLOpeners should not assume that the URL
+// has a particular scheme.
+type WidgetURLOpener interface {
+  OpenWidgetURL(ctx context.Context, u *url.URL) (*Widget, error)
+}
+
+// A type that implements NamedWidgetURLOpener has a preferred scheme for
+// the URLs it opens. WidgetURLScheme must be safe to call from multiple
+// goroutines.
+type NamedWidgetURLOpener interface {
+  WidgetURLOpener
+  WidgetURLScheme() string
+}
+
+// URLMux is a URL opener multiplexer. It matches the scheme of the URLs
+// against a set of registered schemes and calls the opener that matches the
+// URL's scheme.
+//
+// The zero value is a URLMux with no registered schemes.
+type URLMux struct {
+  // ...
+}
+
+// AddWidget registers the opener with its preferred scheme. If an opener
+// already exists for the scheme, AddWidget panics.
+func (mux *URLMux) AddWidget(opener NamedWidgetURLOpener) {
+  mux.RegisterWidget(opener.WidgetURLScheme(), opener)
+}
+
+// RegisterWidget registers the opener with the given scheme. If an opener
+// already exists for the scheme, RegisterWidget panics.
+func (mux *URLMux) RegisterWidget(scheme string, opener WidgetURL) {
+  // ...
+}
+
+// OpenWidget calls OpenWidgetURL with the URL parsed from urlstr.
+// OpenWidget is safe to call from multiple goroutines.
+func (mux *URLMux) OpenWidget(ctx context.Context, urlstr string) (*Widget, error) {
+  u, err := url.Parse(urlstr)
+  if err != nil {
+    return nil, fmt.Errorf("open widget: %v", err)
+  }
+  return mux.OpenWidgetURL(ctx, u)
+}
+
+// OpenWidgetURL dispatches the URL to the opener that is registered with the
+// URL's scheme. OpenWidget is safe to call from multiple goroutines.
+func (mux *URLMux) OpenWidgetURL(ctx context.Context, u *url.URL) (*Widget, error) {
+  // ...
+}
+
+// DefaultURLMux returns the URLMux used by OpenWidget.
+func DefaultURLMux() *URLMux {
+  return defaultURLMux
+}
+
+var defaultURLMux = new(URLMux)
+
+// OpenWidget opens the Widget identified by the URL given. URL openers must be
+// registered in the DefaultURLMux, which is typically done in driver
+// packages' initialization.
+func OpenWidget(ctx context.Context, urlstr string) (*Widget, error) {
+  return DefaultURLMux().OpenWidget(urlstr)
+}
+```
+
+The repetition of `Widget` in the method names permits a type to handle multiple
+resources within the API. Exporting the `URLMux` allows applications to build
+their own muxes, potentially wrapping existing.
+
+Driver packages should include their own `URLOpener` struct type which
```

Comment:
Can you given an example of what a driver implementation would look like? (including an example of setting up the struct to allow credentials to be passed in, as well as the `init` function registering a default implementation of URLOpener for scheme(s)).

Reply:
IMHO it will be useful to include such a snippet as an example. Something like:

```
const Scheme = "gs"

type URLOpener struct {
	Client  *gcp.HTTPClient
	Options Options
}

func (o *URLOpener) OpenBucketURL(ctx context.Context, u *url.URL) (*blob.Bucket, error) {
    ...
	return OpenBucket(ctx, o.Client, u.Host, &o.Options)
}

func init() {
    blob.DefaultURLMux().Register(Scheme, URLOpener{})
}
```

as well as example of the two ways of using it:

```
// Using default mux. Import provider package, then:
bucket, err := blob.DefaultURLMux().OpenURL(ctx, "<url>")

// Custom mux.
myMux := blob.URLMux{}
myMux.Register(gcsblob.Scheme, gcsblob.URLOpener{Client: client})
myMux.Register(fileblob.Scheme, fileblob.URLOpener{})
...
bucket, err := myMux.OpenURL(ctx, "<url>")
```

Reply:
Done.